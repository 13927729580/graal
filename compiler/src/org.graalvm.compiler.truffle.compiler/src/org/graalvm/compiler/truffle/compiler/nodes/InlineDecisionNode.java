/*
 * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package org.graalvm.compiler.truffle.compiler.nodes;

import org.graalvm.compiler.core.common.type.StampFactory;
import org.graalvm.compiler.graph.IterableNodeType;
import org.graalvm.compiler.graph.NodeClass;
import org.graalvm.compiler.nodeinfo.NodeCycles;
import org.graalvm.compiler.nodeinfo.NodeInfo;
import org.graalvm.compiler.nodeinfo.NodeSize;
import org.graalvm.compiler.nodes.ConstantNode;
import org.graalvm.compiler.nodes.ValueNode;

import jdk.vm.ci.meta.JavaKind;

/**
 * Since SVM specializes behaviour on call sites, we need to differentiate the inlined call sites
 * which should not contain this specialized code.
 * 
 * This node is a placeholder to be replaced with a {@code true} or {@code false} constant matching
 * the corresponding inlining decision. The inlining decision is made on the invoke which is
 * connected to this node with data flow through the {@link #handle} i.e. The handle is a
 * {@link InlineDecisionHandleNode} whose other usage is a {@link InlineDecisionAttachNode} wrapping
 * the arguments of the call whose inlining decision is guarded by this node.
 */
@NodeInfo(cycles = NodeCycles.CYCLES_0, size = NodeSize.SIZE_0)
public final class InlineDecisionNode extends ValueNode implements IterableNodeType {

    public static final NodeClass<InlineDecisionNode> TYPE = NodeClass.create(InlineDecisionNode.class);
    // Used by the language agnostic inlining to locate this node from the Invoke
    @SuppressWarnings("unused") @Input private InlineDecisionHandleNode handle;

    protected InlineDecisionNode(InlineDecisionHandleNode handle) {
        super(TYPE, StampFactory.forKind(JavaKind.Boolean));
        this.handle = handle;
    }

    public static InlineDecisionNode create(InlineDecisionHandleNode handle) {
        return new InlineDecisionNode(handle);
    }

    public void inlined() {
        replaceWith(true);
    }

    public void notInlined() {
        replaceWith(false);
    }

    private void replaceWith(boolean b) {
        replaceAtUsagesAndDelete(graph().unique(ConstantNode.forBoolean(b)));
    }
}
