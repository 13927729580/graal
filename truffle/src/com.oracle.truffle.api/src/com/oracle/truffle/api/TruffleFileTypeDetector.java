/*
 * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or
 * data (collectively the "Software"), free of charge and under any and all
 * copyright rights in the Software, and any and all patent rights owned or
 * freely licensable by each licensor hereunder covering either (i) the
 * unmodified Software as contributed to or provided by such licensor, or (ii)
 * the Larger Works (as defined below), to deal in both
 *
 * (a) the Software, and
 *
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 *
 * The above copyright notice and either this complete permission notice or at a
 * minimum a reference to the UPL must be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.oracle.truffle.api;

import java.io.IOException;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.graalvm.polyglot.io.FileSystem;

/**
 * A MIME type detector for finding file's MIME type and encoding.
 *
 * <p>
 * The means by which the detector determines the MIME is highly implementation specific. A simple
 * implementation might detect the MIME type using {@link TruffleFile file} extension. In other
 * cases, the content of {@link TruffleFile file} needs to be examined to guess its file type.
 *
 * @see TruffleFile#getMimeType()
 * @since 1.0
 */
public interface TruffleFileTypeDetector {

    /**
     * Finds a MIME type for given {@link TruffleFile}.
     *
     * @param file the {@link TruffleFile file} to find a MIME type for
     * @return the MIME type or {@code null} if the MIME type is not recognized
     * @throws IOException of an I/O error occurs
     * @throws SecurityException if the implementation requires an access the file and the
     *             {@link FileSystem} denies the operation
     * @since 1.0
     */
    String findMimeType(TruffleFile file) throws IOException;

    /**
     * For a file containing an encoding information returns the encoding.
     *
     * @param file the {@link TruffleFile file} to find an encoding for
     * @return the file encoding or {@code null} if the file does not provide encoding
     * @throws IOException of an I/O error occurs
     * @throws SecurityException if the {@link FileSystem} denies the file access
     * @since 1.0
     */
    String findEncoding(TruffleFile file) throws IOException;

    /**
     * The annotation to use to register {@link TruffleFileTypeDetector}. By annotating your
     * implementation of {@link TruffleFileTypeDetector} by this annotation the detector will be
     * used by {@link TruffleFile}.
     *
     * @since 1.0
     */
    @Retention(RetentionPolicy.CLASS)
    @Target(ElementType.TYPE)
    @interface Registration {
    }
}
