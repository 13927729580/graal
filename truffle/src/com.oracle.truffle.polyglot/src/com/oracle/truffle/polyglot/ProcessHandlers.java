/*
 * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or
 * data (collectively the "Software"), free of charge and under any and all
 * copyright rights in the Software, and any and all patent rights owned or
 * freely licensable by each licensor hereunder covering either (i) the
 * unmodified Software as contributed to or provided by such licensor, or (ii)
 * the Larger Works (as defined below), to deal in both
 *
 * (a) the Software, and
 *
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 *
 * The above copyright notice and either this complete permission notice or at a
 * minimum a reference to the UPL must be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.oracle.truffle.polyglot;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Paths;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.TimeUnit;
import org.graalvm.polyglot.io.ProcessHandler;

final class ProcessHandlers {

    private ProcessHandlers() {
        throw new IllegalStateException("Instance is not allowed");
    }

    static ProcessHandler newDefaultProcessHandler() {
        return new DefaultProcessHandler();
    }

    static boolean isDefault(ProcessHandler handler) {
        return handler instanceof DefaultProcessHandler;
    }

    private static final class DefaultProcessHandler implements ProcessHandler {

        @Override
        public Process start(ProcessCommand command) throws IOException {
            ProcessBuilder builder = new ProcessBuilder(command.getCommand()).redirectErrorStream(command.isRedirectErrorStream()).redirectInput(
                            translateRedirect(command.getInputRedirect())).redirectOutput(translateRedirect(command.getOutputRedirect())).redirectError(translateRedirect(command.getErrorRedirect()));
            Map<String, String> env = builder.environment();
            env.clear();
            env.putAll(command.getEnvironment());
            String cwd = command.getDirectory();
            if (cwd != null) {
                builder.directory(Paths.get(cwd).toFile());
            }
            Process process = builder.start();
            boolean outputRedirectedToStream = command.getOutputRedirect().getType() == ProcessHandler.Redirect.Type.STREAM;
            boolean errorOutputRedirectedToStream = command.getErrorRedirect().getType() == ProcessHandler.Redirect.Type.STREAM;
            if (outputRedirectedToStream || errorOutputRedirectedToStream) {
                process = new ProcessDecorator(
                                command.getCommand().get(0),
                                process,
                                command.getOutputRedirect().getOutputStream(),
                                command.getErrorRedirect().getOutputStream());
            }
            return process;
        }

        private static java.lang.ProcessBuilder.Redirect translateRedirect(ProcessHandler.Redirect redirect) {
            switch (redirect.getType()) {
                case PIPE:
                    return java.lang.ProcessBuilder.Redirect.PIPE;
                case INHERIT:
                    return java.lang.ProcessBuilder.Redirect.INHERIT;
                case STREAM:
                    return java.lang.ProcessBuilder.Redirect.PIPE;
                default:
                    throw new IllegalStateException("Unsupported redirect: " + redirect);
            }
        }

        private static final class ProcessDecorator extends Process {

            private final Process delegate;
            private final Thread outCopier;
            private final Thread errCopier;

            ProcessDecorator(
                            String name,
                            Process delegate,
                            OutputStream out,
                            OutputStream err) {
                Objects.requireNonNull(delegate, "Delegate must be non null.");
                this.delegate = delegate;
                this.outCopier = out == null ? null : new CopierThread(name + " [stdout]", delegate.getInputStream(), out);
                this.errCopier = err == null ? null : new CopierThread(name + " [stderr]", delegate.getErrorStream(), err);
                if (outCopier != null) {
                    outCopier.start();
                }
                if (errCopier != null) {
                    errCopier.start();
                }
            }

            @Override
            public OutputStream getOutputStream() {
                return delegate.getOutputStream();
            }

            @Override
            public InputStream getInputStream() {
                if (outCopier == null) {
                    return delegate.getInputStream();
                }
                return null;
            }

            @Override
            public InputStream getErrorStream() {
                if (errCopier == null) {
                    return delegate.getErrorStream();
                }
                return null;
            }

            @Override
            public int waitFor() throws InterruptedException {
                int res = delegate.waitFor();
                waitForCopiers();
                return res;
            }

            @Override
            public int exitValue() {
                return delegate.exitValue();
            }

            @Override
            public void destroy() {
                delegate.destroy();
            }

            @Override
            public Process destroyForcibly() {
                delegate.destroyForcibly();
                return this;
            }

            @Override
            public boolean isAlive() {
                return delegate.isAlive();
            }

            @Override
            public boolean waitFor(long timeout, TimeUnit unit) throws InterruptedException {
                boolean res = delegate.waitFor(timeout, unit);
                if (res) {
                    waitForCopiers();
                }
                return res;
            }

            private void waitForCopiers() throws InterruptedException {
                if (outCopier != null) {
                    outCopier.join();
                }
                if (errCopier != null) {
                    errCopier.join();
                }
            }
        }

        private static final class CopierThread extends Thread {

            private static final int BUFSIZE = 8192;

            private final InputStream in;
            private final OutputStream out;
            private final byte[] buffer;

            CopierThread(String name, InputStream in, OutputStream out) {
                Objects.requireNonNull(name, "Name must be non null.");
                Objects.requireNonNull(in, "In must be non null.");
                Objects.requireNonNull(out, "Out must be non null.");
                setName(name);
                this.in = in;
                this.out = out;
                this.buffer = new byte[BUFSIZE];
            }

            @Override
            public void run() {
                try {
                    while (true) {
                        if (isInterrupted()) {
                            return;
                        }
                        int read = in.read(buffer, 0, buffer.length);
                        if (read == -1) {
                            return;
                        }
                        out.write(buffer, 0, read);
                    }
                } catch (IOException e) {
                }
            }
        }

    }
}
