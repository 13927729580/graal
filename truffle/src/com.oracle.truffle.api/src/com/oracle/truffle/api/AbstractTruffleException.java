/*
 * Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or
 * data (collectively the "Software"), free of charge and under any and all
 * copyright rights in the Software, and any and all patent rights owned or
 * freely licensable by each licensor hereunder covering either (i) the
 * unmodified Software as contributed to or provided by such licensor, or (ii)
 * the Larger Works (as defined below), to deal in both
 *
 * (a) the Software, and
 *
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 *
 * The above copyright notice and either this complete permission notice or at a
 * minimum a reference to the UPL must be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.oracle.truffle.api;

import com.oracle.truffle.api.TruffleStackTrace.LazyStackTrace;
import com.oracle.truffle.api.nodes.Node;
import com.oracle.truffle.api.source.SourceSection;


@SuppressWarnings("serial")
public abstract class AbstractTruffleException extends RuntimeException implements TruffleException {

    private volatile LazyStackTrace lazy;

    protected AbstractTruffleException() {
        this(null, null);
    }

    protected AbstractTruffleException(String message) {
        this(message, null);
    }

    protected AbstractTruffleException(Throwable throwable) {
        this(null, throwable);
    }

    protected AbstractTruffleException(String message, Throwable throwable) {
        super(message, throwable);
    }

    /**
     * Returns a node indicating the location where this exception occurred in the AST. This method
     * may return <code>null</code> to indicate that the location is not available.
     *
     */
    public abstract Node getLocation();

    /**
     * Returns an additional guest language object. The return object must be an interop type so it
     * must be either implementing TruffleObject or be a primitive value. The default implementation
     * returns <code>null</code> to indicate that no object is available for this exception.
     *
     */
    @Override
    public Object getExceptionObject() {
        return null;
    }

    /**
     * Returns <code>true</code> if this exception indicates a parser or syntax error. Syntax errors
     * typically occur while
     * {@link TruffleLanguage#parse(com.oracle.truffle.api.TruffleLanguage.ParsingRequest) parsing}
     * of guest language source code.
     *
     */
    @Override
    public boolean isSyntaxError() {
        return false;
    }

    /**
     * Returns <code>true</code> if this exception indicates a syntax error that is indicating that
     * the syntax is incomplete. This allows guest language programmers to find out if more code is
     * expected from a given source. For example an incomplete JavaScript program could look like
     * this:
     *
     * <pre>
     * function incompleteFunction(arg) {
     * </pre>
     *
     * A shell might react to this exception and prompt for additional source code, if this method
     * returns <code>true</code>.
     *
     */
    @Override
    public boolean isIncompleteSource() {
        return false;
    }

    /**
     * Returns <code>true</code> if this exception indicates an internal error. Note that all
     * exceptions thrown in a guest language implementation that are not implementing
     * {@link TruffleException} are considered internal.
     *
     * @since 0.27
     */
    @Override
    public boolean isInternalError() {
        return false;
    }

    /**
     * Returns <code>true</code> if this exception indicates that guest language application was
     * cancelled during its execution.
     *
     */
    @Override
    public boolean isCancelled() {
        return false;
    }

    /**
     * Returns <code>true</code> if the exception indicates that the application was exited within
     * the guest language program. If {@link #isExit()} returns <code>true</code> also
     * {@link #getExitStatus()} should be implemented.
     *
     * @see #getExitStatus()
     */
    @Override
    public boolean isExit() {
        return false;
    }

    /**
     * Returns the exit status if this exception indicates that the application was {@link #isExit()
     * exited}. The exit status is intended to be passed to {@link System#exit(int)}.
     *
     * @see #isExit()
     */
    @Override
    public int getExitStatus() {
        return 0;
    }

    /**
     * Returns the number of guest language frames that should be collected for this exception.
     * Returns a negative integer by default for unlimited guest language frames. This is intended
     * to be used by guest languages to limit the number of guest language stack frames. Languages
     * might want to limit the number of frames for performance reasons. Frames that point to
     * {@link RootNode#isInternal() internal} internal root nodes are not counted when the stack
     * trace limit is computed.
     *
     */
    @Override
    public int getStackTraceElementLimit() {
        return -1;
    }

    /**
     * Returns a location where this exception occurred in the AST. This method may return
     * <code>null</code> to indicate that the location is not available.
     *
     * @return the {@link SourceSection} or null
     */
    @Override
    public SourceSection getSourceLocation() {
        final Node node = getLocation();
        return node == null ? null : node.getEncapsulatingSourceSection();
    }

    LazyStackTrace getLazyStackTrace() {
        LazyStackTrace res = lazy;
        if (res == null) {
            synchronized (this) {
                res = lazy;
                if (res == null) {
                    res = new LazyStackTrace();
                    lazy = res;
                }
            }
        }
        return res;
    }
}
