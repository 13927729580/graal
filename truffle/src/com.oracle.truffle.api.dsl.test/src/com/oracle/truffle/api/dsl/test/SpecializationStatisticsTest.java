/*
 * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or
 * data (collectively the "Software"), free of charge and under any and all
 * copyright rights in the Software, and any and all patent rights owned or
 * freely licensable by each licensor hereunder covering either (i) the
 * unmodified Software as contributed to or provided by such licensor, or (ii)
 * the Larger Works (as defined below), to deal in both
 *
 * (a) the Software, and
 *
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 *
 * The above copyright notice and either this complete permission notice or at a
 * minimum a reference to the UPL must be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.oracle.truffle.api.dsl.test;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.net.URL;
import java.net.URLConnection;
import java.nio.charset.StandardCharsets;
import java.util.stream.Collectors;

import org.graalvm.polyglot.Context;
import org.junit.Assert;
import org.junit.Test;

import com.oracle.truffle.api.Assumption;
import com.oracle.truffle.api.Truffle;
import com.oracle.truffle.api.dsl.Cached;
import com.oracle.truffle.api.dsl.GenerateUncached;
import com.oracle.truffle.api.dsl.NodeField;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.dsl.SpecializationStatistics;
import com.oracle.truffle.api.dsl.test.SpecializationStatisticsTestFactory.SpecializationStatisticTestNodeGen;
import com.oracle.truffle.api.nodes.Node;
import com.oracle.truffle.api.source.Source;
import com.oracle.truffle.api.source.SourceSection;

public class SpecializationStatisticsTest {

    @GenerateUncached
    @NodeField(name = "index", type = int.class)
    @SpecializationStatistics.AlwaysEnabled
    abstract static class SpecializationStatisticTestNode extends Node {

        abstract Object execute(Object arg);

        @Specialization(guards = "arg == cachedArg", limit = "3", assumptions = "createAssumption()")
        public int s0(@SuppressWarnings("unused") int arg, @Cached("arg") int cachedArg) {
            return cachedArg;
        }

        @Specialization(replaces = "s0")
        public int s1(@SuppressWarnings("unused") int arg) {
            return arg;
        }

        protected abstract int getIndex();

        static Assumption createAssumption() {
            return Truffle.getRuntime().createAssumption();
        }

        @Specialization
        public String s2(String arg) {
            return arg;
        }

        @Specialization(replaces = "s2")
        public CharSequence s3(CharSequence arg) {
            return arg;
        }

        @Override
        public SourceSection getSourceSection() {
            if (!isAdoptable()) {
                return null;
            }
            return Source.newBuilder("testLang", String.valueOf(getIndex()), "testLangFile" + getIndex() + ".file").build().createSection(1);
        }

    }

    @Test
    public void testCustomEnterLeave() throws IOException {
        SpecializationStatistics statistics = SpecializationStatistics.create();
        SpecializationStatistics prev = statistics.enter();

        createAndExecuteNodes();

        StringWriter writer = new StringWriter();
        statistics.printHistogram(new PrintWriter(writer));

        String contents = readExpectedOutput();

        Assert.assertEquals(contents, writer.toString());

        statistics.leave(prev);
    }

    private static String readExpectedOutput() throws IOException {
        URL url = SpecializationStatisticsTest.class.getResource("SpecializationStatisticsTest.out");
        String contents;
        URLConnection conn = url.openConnection();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8))) {
            contents = reader.lines().collect(Collectors.joining(System.lineSeparator()));
        }
        return contents;
    }

    private static void createAndExecuteNodes() {
        SpecializationStatisticTestNode node = SpecializationStatisticTestNodeGen.create(0);
        node.execute(42);
        node.execute(43);
        node.execute(44);
        node.execute(45);
        node.execute("");
        node.execute(new StringBuilder());
        node.execute("");

        node = SpecializationStatisticTestNodeGen.create(1);
        node.execute("");
        node.execute("");
        node.execute("");

        node = SpecializationStatisticTestNodeGen.getUncached();
        node.execute("");
    }

    @Test
    public void testWithContextEnabled() throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (Context context = Context.newBuilder().allowExperimentalOptions(true).option("engine.SpecializationStatistics", "true").logHandler(out).build()) {
            context.enter();
            createAndExecuteNodes();
            context.leave();
        }
        Assert.assertEquals(createLogEntry(readExpectedOutput()), new String(out.toByteArray()));
    }

    private static String createLogEntry(String logMessage) {
        return String.format("[engine] Specialization histogram: %n%s%n", logMessage);
    }

    @Test
    public void testWithContextDisabled() {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (Context context = Context.newBuilder().allowExperimentalOptions(true).option("engine.SpecializationStatistics", "false").logHandler(out).build()) {
            context.enter();
            createAndExecuteNodes();
            context.leave();
        }
        Assert.assertEquals("", new String(out.toByteArray()));
    }

    @Test
    public void testWithContextNoExecute() {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (Context context = Context.newBuilder().allowExperimentalOptions(true).option("engine.SpecializationStatistics", "true").logHandler(out).build()) {
            context.enter();
            context.leave();
        }
        Assert.assertEquals(createLogEntry("No specialization statistics data was collected. " +
                        "Either no node with @Specialization annotations was executed or the interpreter was not compiled with -J-Dtruffle.dsl.GenerateSpecializationStatistics=true e.g as parameter to the javac tool."),
                        new String(out.toByteArray()));
    }

}
