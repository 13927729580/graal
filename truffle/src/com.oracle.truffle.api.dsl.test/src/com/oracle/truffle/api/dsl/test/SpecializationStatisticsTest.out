Name                                                  Instances          Executions     Executions per instance 
  SpecializationStatisticTestNodeGen.Uncached         1 (33%)             1 (9%)         Min=         1 Avg=        1.00 Max          1  MaxNode N/A 
    s0                                                  0 (0%)              0 (0%)         Min=         0 Avg=        0.00 Max          0  MaxNode  -  
    s1                                                  0 (0%)              0 (0%)         Min=         0 Avg=        0.00 Max          0  MaxNode  -  
    s2                                                  0 (0%)              0 (0%)         Min=         0 Avg=        0.00 Max          0  MaxNode  -  
    s3 <String>                                         1 (100%)            1 (100%)       Min=         1 Avg=        1.00 Max          1  MaxNode N/A 
Name                                                  Instances          Executions     Executions per instance 
  SpecializationStatisticTestNodeGen                  2 (67%)            10 (91%)        Min=         3 Avg=        5.00 Max          7  MaxNode testLangFile0.file~1:0 
    s0 <int>                                            1 (50%)             3 (30%)        Min=         3 Avg=        3.00 Max          3  MaxNode testLangFile0.file~1:0 
    s1 <int>                                            1 (50%)             1 (10%)        Min=         1 Avg=        1.00 Max          1  MaxNode testLangFile0.file~1:0 
    s2 <String>                                         2 (100%)            4 (40%)        Min=         1 Avg=        2.00 Max          3  MaxNode testLangFile1.file~1:0 
    s3                                                  1 (50%)             2 (20%)        Min=         2 Avg=        2.00 Max          2  MaxNode testLangFile0.file~1:0 
      <String>                                            1 (100%)            1 (50%)        Min=         1 Avg=        1.00 Max          1  MaxNode testLangFile0.file~1:0 
      <StringBuilder>                                     1 (100%)            1 (50%)        Min=         1 Avg=        1.00 Max          1  MaxNode testLangFile0.file~1:0 

